### PR TITLE
Optimized the slicing part

### PR DESCRIPTION
**Pull request by marcus-bigrep**
changed the main loop from faces/layers to layers/faces
added some prechecks
parallization

### CODE REVIEW

**Comment by Ghostkeeper** [comment]
We've made the main loop explicitly as faces -> layers as an optimisation. One face spans multiple layers, and it's clear which layers it spans; just look at the Z coordinates of all its vertices. So you essentially go through all faces once (which are many) and for each face you iterate through all its layers, which are often like 2 or 3 (or even 0 layers if the triangle is horizontal). The other way around you'd need to go through all layers once and for each layer go through all faces, which is much more work.

The other changes sound good to me, although parallelisation is maybe not desired since the current solution really only takes a fraction of a second for any normal print. Prechecks are nice of course and reducing pointer look-up should give a bit of a performance boost.

Do you have performance measurements to back up your claim that it's been improved?

I haven't done an actual code review yet, mind.

**Comment by marcus-bigrep** [comment]
For the samples, which have a significant percentage of slicing, I have (new/old):
- Just slicing part: 0.37
- Entire run: ‭0.91

**Comment by marcus-bigrep** [comment]
Sorry, I was referring to my own samples. We have a pretty big printer (1 cubic meter).
10% of my samples have a significant percentage in the slicing part. 
These 10% of samples have about a 10% reduction in entire slicing time.
(BTW, my average is 3.5% instead of 1.85% which is yours).

File: src/slicer.cpp [hunk 1]
```diff
@@ -787,72 +787,59 @@ void SlicerLayer::makePolygons(const Mesh* mesh)
...
+}
 
-        // calculate all intersections between a layer plane and a triangle
-        for (unsigned int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
+void Slicer::buildSegments(const Mesh &mesh, const std::vector<std::pair<int32_t, int32_t>> &zbbox,
+    std::vector<SlicerLayer>& layers) 
+{
+    // OpenMP
+#pragma omp parallel for default(none) shared(mesh, zbbox, layers)
+    for (int layer_nr = 0; layer_nr < layers.size(); layer_nr++)
```
Comment (Ghostkeeper) [inline src/slicer.cpp]: Comparing integers of different signedness (`Wsign-compare` in GCC).

File: src/slicer.cpp [hunk 2]
```diff
@@ -787,72 +787,59 @@ void SlicerLayer::makePolygons(const Mesh* mesh)
...
+    {
+        int32_t z = layers[layer_nr].z;
+        layers[layer_nr].segments.reserve(100);
+
+        // loop over all mesh faces
+        for (unsigned int mesh_idx = 0; mesh_idx < mesh.faces.size(); mesh_idx++)
         {
-            int32_t z = layers.at(layer_nr).z;
+            if ((z < zbbox[mesh_idx].first) || (z > zbbox[mesh_idx].second))
+                 continue;
```
Comment (Ghostkeeper) [inline src/slicer.cpp]: So you've now reversed this for loop from `faces -> layers` to `layers -> faces`. This requires you to build this `zbbox` mapping though in order to limit the layers that it's iterating over for each face. I don't see the point of this change; when the original loop nesting was used, it could keep the minimum and maximum Z coordinates local in the loop so it doesn't need to track them in a separate mapping.

File: src/slicer.cpp (hunk 1 above)
Comment (Ghostkeeper) [reply src/slicer.cpp]: I think the best option here is a `LayerIndex` because it's very semantic, but then you'd need a cast to tell the compiler that you're working with layer indices here because that vector doesn't return `LayerInd`ices.

File: src/slicer.cpp [hunk 3]
```diff
@@ -962,27 +987,50 @@ Slicer::Slicer(Mesh* mesh, const coord_t thickness, const size_t slice_layer_cou
...
+        Point3 p2 = v2.p;
+
+        // find the minimum and maximum z point		
+        int32_t minZ = p0.z;
+        if (p1.z < minZ) minZ = p1.z;
+        if (p2.z < minZ) minZ = p2.z;
+
+        int32_t maxZ = p0.z;
+        if (p1.z > maxZ) maxZ = p1.z;
+        if (p2.z > maxZ) maxZ = p2.z;
```
Comment (Ghostkeeper) [inline src/slicer.cpp]: Code style: Brackets on new line.

File: src/slicer.cpp [hunk 4]
```diff
@@ -962,27 +987,50 @@ Slicer::Slicer(Mesh* mesh, const coord_t thickness, const size_t slice_layer_cou
...
+
+        zHeights.emplace_back(std::make_pair(minZ, maxZ));
+    }
+
+    return zHeights;
 }
 
-SlicerSegment Slicer::project2D(const Point3& p0, const Point3& p1, const Point3& p2, const coord_t z) const
```
Comment (Ghostkeeper) [inline src/slicer.cpp]: Why did you remove this `const`?

File: src/slicer.h [hunk 5]
```diff
@@ -504,7 +508,13 @@ class Slicer
...
      * \return The Y coordinate of the point to find.
      */
-    coord_t interpolate(const coord_t x, const coord_t x0, const coord_t x1, const coord_t y0, const coord_t y1) const;
+    static coord_t interpolate(const coord_t x, const coord_t x0, const coord_t x1, const coord_t y0, const coord_t y1)
+    {
+        const coord_t dx_01 = x1 - x0;
+        coord_t num = (y1 - y0) * (x - x0);
+        num += num > 0 ? dx_01 / 2 : -dx_01 / 2; // add in offset to round result
+        return y0 + num / dx_01;
+    }
```
Comment (Ghostkeeper) [inline src/slicer.h]: Code style: No implementation in headers.

File: src/slicer.h [hunk 6]
```diff
@@ -518,9 +528,23 @@ class Slicer
...
+
+    /*! Create the polygons in layers. */
+    static void makePolygons(Mesh &mesh, SlicingTolerance slicing_tolerance, std::vector<SlicerLayer>& layers);
+
+    /* Creates a vector of layers and set their z value. */
+    static std::vector<SlicerLayer> buildLayersWithHeight(size_t slice_layer_count, SlicingTolerance slicing_tolerance,
+        coord_t initial_layer_thickness, coord_t thickness, bool use_variable_layer_heights,
+        const std::vector<AdaptiveLayer> *adaptive_layers);
+
+    /* Creates the segments and write them into the layers. */
```
Comment (Ghostkeeper) [inline src/slicer.h]: Could you write these documentation entries in Doxygen's style?

File: src/slicer.cpp (hunk 2 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: Enabling parallelization is the main point about reversing the loop.

File: src/slicer.cpp (hunk 3 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: Do you a style guide where I can look into?

File: src/slicer.cpp (hunk 4 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: I replaced "const" by "static". This is slightly more efficient.

File: src/slicer.cpp (hunk 1 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: I am afraid, OpenMP is not compatible with LayerIndex (at lease for MSVC).
BTW, what do you think about the new standard "std::for_each(std::execution::par,...)"?"

File: src/slicer.cpp (hunk 3 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: Fixed

File: src/slicer.h (hunk 6 above)
Comment (marcus-bigrep) [reply src/slicer.h]: fixed

File: src/slicer.cpp (hunk 2 above)
Comment (Ghostkeeper) [reply src/slicer.cpp]: Why couldn't the parallelisation work over the original faces loop, then?

File: src/slicer.cpp (hunk 3 above)
Comment (Ghostkeeper) [reply src/slicer.cpp]: The code style guidelines are here: https://github.com/Ultimaker/Meta

File: src/slicer.cpp (hunk 1 above)
Comment (Ghostkeeper) [reply src/slicer.cpp]: We compile CuraEngine with GCC though, not MSVC. Nice to hear though that MSVC is still working (for the rest then). Alternatively you can use int and cast the size to an int, of course. I can't guarantee that we won't give a for loop a LayerIndex in the future though, since we don't test (or support) MSVC.

I think that `std::for_each` with parallel execution is a very nice alternative, if it works for us. It'd also work with MacOS, which is as of now still single-threaded. I hope it works fine together with OpenMP so that it can be transitioned softly rather than having to make a hard atomic transition.
Currently we're still compiling with C++11 though. Hopefully soon that will change to C++17.

File: src/slicer.cpp (hunk 2 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: Because you need write access on the layers. Thus it does not make much sense to parallize the faces loop. Because then the layer part must be "critical". 
In particular, the layer loop shall be "heavy" to boost up parallisation. Thus I have reversed the loops. Otherwise you parallize just over a few layers.

File: src/slicer.cpp (hunk 1 above)
Comment (marcus-bigrep) [reply src/slicer.cpp]: Apart from GCodeLayerThreader, CuraEngine parallization is just about loops and 
these loops are executed "sequentially". So the migration ought to be quite smooth. Maybe even GCodeLayerThreader could be converted to a simple loop...
Is there a specific time line to upgrade to C++17?

Comment (Ghostkeeper) [reply src/slicer.cpp]: Not really since we don't generally work with timelines and promises to fix something before a certain deadline. It was recently brought up again and moved up a bit in our backlog. Most likely we'd plan this in after the 4.5 branch has split off to reduce risk for that release (though personally I won't think there's any risk at all).

Comment (marcus-bigrep) [reply src/slicer.cpp]: Thanks. I am looking forward!

**Comment by marcus-bigrep** [comment]
Do you need more information?

**Comment by Ghostkeeper** [comment]
No, sorry, other things got in between for a bit. We'll review the latest changes again.

File: src/slicer.cpp (hunk 2 above)
Comment (Ghostkeeper) [reply src/slicer.cpp]: Ah yes, I see. That's a good reason.

**Comment by Ghostkeeper** [comment]
The slice result is still the same as far as I've found. I also found a performance improvement of 78% on average of the slicing stage. Good job!

**Comment by jdumas** [comment]
Hi. I was reading the [documentation](https://github.com/Ultimaker/CuraEngine/blob/master/docs/slicing.md) that says you iterate over faces/layers, and was surprised to find the implementation iterates over layers/faces instead. Have you tried slicing *really* large meshes (+10M faces at least)? Unless you have a lot of cores, a serial implementation of the loop over faces/layers should be faster still. Even then, it is possible to make it parallel using `tbb::concurrent_vector<>` for multi-threaded segment insertions. I found that I can slice a mesh with 67M faces in about 3s on a laptop using that approach, which is really fast.

**Comment by marcus-bigrep** [comment]
@jdumas Thanks for the remark. I'll adjust the docu.
I did benchmarks with a lot of big stls and it is much faster. The timings have been discussed in this pull request in detail. In particular, we are more than 2x faster for the slicing part (actually a lot more because most time is consumed in the stitching now which is part of the duration measurement).
We build a precheck structure which we test against. This is our way to integrate the optimization idea of the original loop order.
tbb::concurrent_vector<> : I have not tested this data structure here but I worked with it before and it has some overhead. I am pretty sure that our approach is better. BTW, some discussions about the ideas are here as well: https://github.com/Ultimaker/CuraEngine/pull/1174#discussion_r368581986

**Comment by jdumas** [comment]
Well, I am not convinced that this approach is faster, but I would be willing to do some comparisons when I have some time. Could you point me out to the relevant performance tests that you have run so that I may try it out? It may take me some time before I get down to it, but it'd be interesting to compare.

**Comment by marcus-bigrep** [comment]
Do it!
I am afraid, but "my" stls are under NDA and I cannot share them. Though I guess that they are not special in terms of this algorithm. So I suggest: try something out with your stls and if numbers are good, I can run it on my stls and share the numbers.

**Comment by Ghostkeeper** [comment]
I've been trying this with rather simple shapes generated using OpenSCAD. In particular these two from my folder of stress tests:

* https://github.com/Ghostkeeper/SlicerTestModels/blob/master/stress_test/cylinder_hi.scad
* https://github.com/Ghostkeeper/SlicerTestModels/blob/master/stress_test/sphere.scad

And this one where I manually increased the resolution:

* https://github.com/Ghostkeeper/SlicerTestModels/blob/master/polytest/support_strength_infill.scad

Be careful with these though. They can easily generate files upwards of 2GB. Especially the last one has a tendency to crash OpenSCAD.

**Comment by Ghostkeeper** [comment]
I've merged Marcus' updates to the docs here: https://github.com/Ultimaker/CuraEngine/pull/1212

Hopefully that makes it more clear.
